lisa/analysis/idle.py

```diff
@@ -72,6 +73,50 @@ def signal_cpu_active(self, cpu):
         # Fix sequences of wakeup/sleep events reported with the same index
         return handle_duplicate_index(cpu_active)
 
+    @memoized
+    @signal_cpu_active.used_events
+    @TasksAnalysis.df_tasks_states.used_events
+    def signal_cpu_sched_active(self, cpu):
+        """
+        Build a square wave representing the active (i.e. non-idle) CPU time
+        from the scheduler's point of view.
+
+        :param cpu: CPU ID
+        :type cpu: int
+
+        :returns: A :class:`pandas.Series` that equals True at timestamps where the
+          CPU is reported to be non-idle, False otherwise
+        """
+        # Find the earliest time at which the scheduler placed a task on
+        # this CPU's rq. The scheduler won't consider this CPU idle at this
+        # time. A proper rq depth signal would make this more accurate.
+        signal_active = self.signal_cpu_active(cpu).copy()
+        df_states = self.trace.analysis.tasks.df_tasks_states()
+
+        df_states = df_states[df_states.target_cpu == cpu]
+        sched_active = pd.Series(index=df_states.index, data=1)
+
+        # If there's an enqueue when the CPU is deemed inactive, we want to
+        # "move" the active signal earlier in the df.
+        # To do that, first set the signal to active whenever there's such a
+        # wakeup, then clear up redundant values.
+        #
+        # XXX: bitwise 'or' has surprising results:
+        # s1 = pd.Series(index=[1.25], data=[True])
+        # s2 = pd.Series(index=[1.05], data=[True])
+        # print(s1 | s2)
+        # 1.05    False
+        # 1.25     True
+        def orify(a, b):
+            return a or b
+
+        signal_active = signal_active.combine(sched_active, orify, fill_value=False)
```

Comment:
might be easier to add all the columns into a dataframe, and then just return a final column ? That ensures indexes are aligned and is generally easier to printf-debug

Reply:
I don't see how to make this simpler using dataframes since I just want to 'or' two signals. Is there something specific you were thinking of?

Reply:
Actually adding the series as a column would give a column of mostly NaN so it's not what we want.

If I understand correctly the current code, it's equivalent to a reindex. The use of `X | 1` is a way of getting `X` on indices of the 1's series and NaN the rest of the time, and then we drop the NaN.

```
df_states = self.trace.analysis.tasks.df_tasks_states()
df_states = df_states[df_states.target_cpu == cpu] 
cpu_active = self.signal_cpu_active(cpu)

active = cpu_active.reindex(df_states.index, method='ffill')
return series_deduplicate(active, consecutives=True, keep='first')
```

Reply:
EDIT: I misunderstood the actual meaning of `fill_value`, i.e. it's replacing NaN in the inputs to the function, not a blanket replace in the output when any of the input is NaN.